lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1550,6 +1557,86 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
   KPE->resolveComponents(TC.Context, components);
 }
 
+Expr *PreCheckExpression::simplifyTypeConstructionWithLiteralArg(Expr *E) {
+  // If constructor call is followed by `BOE` or `FVE` it means that such
+  // call is supposed to produce optional and literal constructors aren't
+  // failable so let's not attempt this optimization.
+  if (!ExprStack.empty()) {
+    auto *parent = ExprStack.back();
+    if (isa<BindOptionalExpr>(parent) || isa<ForceValueExpr>(parent))
+      return nullptr;
+  }
+
+  auto *call = dyn_cast<CallExpr>(E);
+  if (!call || call->getNumArguments() != 1)
+    return nullptr;
+
+  auto *typeExpr = dyn_cast<TypeExpr>(call->getFn());
+  if (!typeExpr)
+    return nullptr;
+
+  auto *argExpr = call->getArg()->getSemanticsProvidingExpr();
+  auto *literal = dyn_cast<LiteralExpr>(argExpr);
+  if (!literal)
+    return nullptr;
+
+  auto *protocol = TC.getLiteralProtocol(literal);
+  if (!protocol)
+    return nullptr;
+
+  Type type;
+  if (auto *rep = typeExpr->getTypeRepr()) {
+    TypeResolutionOptions options;
+    options |= TypeResolutionFlags::AllowUnboundGenerics;
+    options |= TypeResolutionFlags::InExpression;
+    type = TC.resolveType(rep, DC, options);
+  } else {
+    type = typeExpr->getTypeLoc().getType();
+  }
+
+  if (!type)
+    return nullptr;
+
+  ConformanceCheckOptions options;
+  options |= ConformanceCheckFlags::InExpression;
+  options |= ConformanceCheckFlags::SkipConditionalRequirements;
+
+  auto result = TC.conformsToProtocol(type, protocol, DC, options);
+  if (!result || !result->isConcrete())
+    return nullptr;
+
+  auto initializers = protocol->lookupDirect(DeclBaseName::createConstructor());
+  // We are only interested in protocol requirements.
+  initializers.erase(std::remove_if(initializers.begin(), initializers.end(),
+                                    [&](ValueDecl *init) {
+                                      return !isa<ProtocolDecl>(
+                                          init->getDeclContext());
```

Comment:
I don't quite follow what you're doing here. What if the init is defined an an extension rather than directly in the protocol?

Reply:
I'm trying to figure out how many actual `init` requirements does particular literal protocol have, if there are more than one, we can't apply this transformation, at the same time something which is declared in the extension is not a requirement, so I'm just filtering such cases out.

Reply:
It does not but I'm not sure what we are going to do in that hypothetical case, because such initializers can have different labels and we don't want to re-write the expression to be a coercion.

Reply:
Can you update the proposal to point this issue out?

We already bake-in knowledge of these protocols in the compiler, it may make sense to also bake-in the specific knowledge of which initializer to call. I haven't really given it much thought, though - it's worth a little discussion in the proposal.

Reply:
I'm not even sure if this approach is better than my original PR but at least it doesn't loose any source information, I tried to see if this could be modeled in constraint system itself but I don't think it's possible because of label in the literal constructor, and we can't simply turn member lookup into coercion disjunction either because we don't know what is argument going to be since it's modeled separately.